### PR TITLE
npmd: delete binary and version files if agent misses capability

### PR DESCRIPTION
**Pull request by chparimi**
If NPMD agent binary does not have raw socket capability then
do not run the agent and delete the version files as well as the
agent binary. DSC in the next poll interval will update the
binary with the proper capabilities.

### CODE REVIEW

**Comment by chparimi** [comment]
This is to ensure that even if upgrade step is utilized by a customer (because of which binary loses the prior assigned capability) the binary will not run and eventually DSC will update the binary with proper capabilities.

**Comment by chparimi** [comment]
@Microsoft/omsagent-devs

File: source/code/plugins/in_npmd_server.rb [hunk 1]
```diff
@@ -334,6 +343,28 @@ def check_and_update_binaries
             File.unlink(_x64BinPath) if File.exist?(_x64BinPath)
         end
 
+        def check_agent_capability
+            return if !File.exist?(@location_agent_binary) or @skip_capability_check
```
Comment (varastog) [inline source/code/plugins/in_npmd_server.rb]: what are we using @skip_capability_check for?

File: source/code/plugins/in_npmd_server.rb [hunk 2]
```diff
@@ -334,6 +343,28 @@ def check_and_update_binaries
...
+
+            # Checking if capabilties are proper else delete binary and version file
+            _getCapResult = `getcap #{@location_agent_binary}`
+            if !_getCapResult.include?(NPMD_AGENT_CAPABILITY)
+                # Delete agent binary
+                File.unlink(@location_agent_binary)
+
+                # Delete version file(s)
+                _fileLocPostFix = "#{NPMD_VERSION_FILE_DIR}/#{NPMD_VERSION_FILE_NAME}"
```
Comment (varastog) [inline source/code/plugins/in_npmd_server.rb]: For path manipulations, wouldn't it be better to use any Path apis?

Comment (chparimi) [inline source/code/plugins/in_npmd_server.rb]: Removing the same.

Also adding the case when NPMD binary is absent but version files are present

File: source/code/plugins/in_npmd_server.rb [hunk 3]
```diff
@@ -79,12 +81,19 @@ def configure(conf)
...
+
         def start
             @binary_presence_test_string = "npmd_agent" if @binary_presence_test_string.nil?
             @binary_invocation_cmd = @location_agent_binary if @binary_invocation_cmd.nil?
+            @npmd_version_dir_prefix = "/var/opt/microsoft/omsagent" if @npmd_version_dir_prefix.nil?
             kill_all_agent_instances()
             upload_pending_stderrors()
             check_and_update_binaries()
+            @do_capability_check = true unless @do_capability_check == false
+            check_agent_capability() unless !@do_capability_check
```
Comment (lagalbra) [inline source/code/plugins/in_npmd_server.rb]: For consistency, I would suggest using the same syntax to check `@do_capability_check`: either `@do_capability_check == false` or `!@do_capability_check`

File: source/code/plugins/in_npmd_server.rb [hunk 4]
```diff
@@ -334,6 +343,34 @@ def check_and_update_binaries
             File.unlink(_x64BinPath) if File.exist?(_x64BinPath)
         end
 
+        def check_agent_capability
+            _deleteVersionFiles = false
+
+            if !File.exist?(@location_agent_binary)
+                _deleteVersionFiles = true
+            else
+                _getCapResult = `getcap #{@location_agent_binary}`
```
Comment (lagalbra) [inline source/code/plugins/in_npmd_server.rb]: Have you verified that the `getcap` program exists on [all distros/versions we support](https://github.com/Microsoft/OMS-Agent-for-Linux/tree/master#supported-linux-operating-systems)?

File: source/code/plugins/in_npmd_server.rb [hunk 5]
```diff
@@ -334,6 +343,34 @@ def check_and_update_binaries
...
+                end
+            end
+
+            if _deleteVersionFiles
+                # Single/First workspace scenario
+                _fileLocPostFix = "#{NPMD_VERSION_FILE_DIR}/#{NPMD_VERSION_FILE_NAME}"
+                _loc = "#{@npmd_version_dir_prefix}/#{_fileLocPostFix}"
+                File.unlink(_loc) if File.exist?(_loc)
+                # Multiple workspace scenario
+                _stateDirs = Dir.glob("#{@npmd_version_dir_prefix}/**/#{NPMD_VERSION_FILE_DIR}")
```
Comment (lagalbra) [inline source/code/plugins/in_npmd_server.rb]: Have you tested this plugin on a machine with multiple workspaces onboarded, and that all version files get deleted?

Comment (chparimi) [inline source/code/plugins/in_npmd_server.rb]: have added that in the unit test.

File: source/code/plugins/in_npmd_server.rb (hunk 3 above)
Comment (chparimi) [inline source/code/plugins/in_npmd_server.rb]: Done

File: source/code/plugins/in_npmd_server.rb (hunk 5 above)
Comment (chparimi) [inline source/code/plugins/in_npmd_server.rb]: The DSC module is not yet ready to add version files for all workspaces. However if files are present this is deleting properly.

**Comment by chparimi** [comment]
Pbuild succeeded as well.

**Pull request merged**